src/replit_river/v2/session.py

```diff
@@ -0,0 +1,1280 @@
+                ) from e
+
+            logger.debug("river client get handshake response : %r", handshake_response)
+            if not handshake_response.status.ok:
+                if handshake_response.status.code == ERROR_CODE_SESSION_STATE_MISMATCH:
+                    do_close()
+
+                raise RiverException(
+                    ERROR_HANDSHAKE,
+                    f"Handshake failed with code {handshake_response.status.code}: {
+                        handshake_response.status.reason
+                    }",
+                )
+
+            # We did it! We're connected!
+            last_error = None
+            rate_limiter.start_restoring_budget(client_id)
+            transition_connected(ws)
+            break
+        except Exception as e:
+            if ws:
+                close_ws_in_background(ws)
+                ws = None
+            last_error = e
+            backoff_time = rate_limiter.get_backoff_ms(client_id)
+            logger.exception(
+                f"Error connecting, retrying with {backoff_time}ms backoff"
+            )
+            await asyncio.sleep(backoff_time / 1000)
+
+    finalize_attempt()
+
+    if last_error is not None:
+        raise RiverException(
+            ERROR_HANDSHAKE,
+            f"Failed to create ws after retrying {max_retry} number of times",
+        ) from last_error
+
+    return True
+
+
+async def _setup_heartbeat(
+    block_until_connected: Callable[[], Awaitable[None]],
+    session_id: str,
+    heartbeat_ms: float,
+    heartbeats_until_dead: int,
+    get_state: Callable[[], SessionState],
+    get_closing_grace_period: Callable[[], float | None],
+    close_websocket: Callable[[], Awaitable[None]],
+    increment_and_get_heartbeat_misses: Callable[[], int],
+) -> None:
+    while True:
+        while (state := get_state()) in ConnectingStates:
+            await block_until_connected()
+
+        if state in TerminalStates:
+            logger.debug(
+                "Session is closed, no need to send heartbeat, state : "
+                "%r close_session_after_this: %r",
+                {state},
+                {get_closing_grace_period()},
+            )
+            # session is closing / closed, no need to send heartbeat anymore
+            break
+
+        await asyncio.sleep(heartbeat_ms / 1000)
+        state = get_state()
+        if state in ConnectingStates:
+            logger.debug("Websocket is not connected, don't expect heartbeat")
+            continue
+
+        if increment_and_get_heartbeat_misses() > heartbeats_until_dead:
+            if get_closing_grace_period() is not None:
+                # already in grace period, no need to set again
+                continue
+            logger.info(
+                "%r closing websocket because of heartbeat misses",
+                session_id,
+            )
+            await close_websocket()
+
+
+async def _serve(
```

Comment:
just out of curiosity, why not a method instead of passing things in?

Reply:
easier to reason about individual functions that have discrete interfaces and conform to their exposed interface. The functions that do the work should further be composed of known quantities.